ufo2fontir/src/source.rs

```diff
@@ -110,43 +130,107 @@ impl Source for DesignSpaceIrSource {
         // UFO filename => map of layer
         let mut layer_cache = HashMap::new();
 
+        let mut glif_locations: HashMap<PathBuf, Vec<DesignSpaceLocation>> = HashMap::new();
+
         for source in designspace.sources {
             // Track files within each UFO
             // The UFO dir *must* exist since we were able to find fontinfo in it earlier
-            let ufo_dir = designspace_dir.join(&source.filename);
+            let ufo_dir = self.designspace_dir.join(&source.filename);
+
+            // TODO less dumbed down version of dim => location mapping
+            let mut location = DesignSpaceLocation::new();
+            source.location.iter().for_each(|dim| {
+                location.insert(dim.name.clone(), dim.xvalue.map(|v| v as i32).unwrap());
+            });
 
             for (glyph_name, glif_file) in glif_files(&ufo_dir, &mut layer_cache, &source)? {
                 if !glif_file.exists() {
                     return Err(Error::FileExpected(glif_file));
                 }
                 let glif_file = glif_file.clone();
                 glyphs.entry(glyph_name).or_default().insert(&glif_file)?;
+
+                let glif_locations = glif_locations.entry(glif_file).or_default();
+                glif_locations.push(location.clone());
             }
         }
 
+        self.glif_locations = (font_info.clone(), glif_locations);
+
         Ok(Input { font_info, glyphs })
     }
 
     fn create_glyph_ir_work(
+        &self,
+        glyph_names: &HashSet<&str>,
+        input: &Input,
+    ) -> Result<Vec<Box<dyn Work<()>>>, Error> {
+        // A single glif could be used by many source blocks that use the same layer
+        // *gasp*
+        // So resolve each file to 1..N locations in designspace
+
+        let mut work: Vec<Box<dyn Work<()>>> = Vec::new();
+
+        // Do we have the location of glifs written down?
+        // TODO: consider just recomputing here instead of failing
+        if input.font_info != self.glif_locations.0 {
```

Comment:
I find a bit opaque the use of an unnamed tuple for this cache of glif locations, when you access things with self.glif_locations.0 and self.glif_locations.1 I wonder what that means. Why do those two have to go together as a tuple, can't they be distinct field of DesignSpaceIrSource struct?
or maybe it's the name of the self.glif_locations that I find confusing; it is appropriate for its .1 item (the hashmap of glif locations), but perhaps not for the tuple as a whole..

Reply:
Clearer?

Reply:
yes thanks